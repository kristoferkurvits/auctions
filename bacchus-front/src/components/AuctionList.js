import React from "react";
import { List } from "semantic-ui-react";
import { AuctionItem } from "./AuctionItem/AuctionItem";


const AuctionList = (props) => {

  return (
    <List divided style={{ maxWidth: 900, margin: "0 auto" }}>
      {props.auctions.map((auction) => (
        <AuctionItem auction={auction} selectedAuctionItemHandler={props.selectedAuctionItemHandler} key={auction.productId} selectedProductId={props.selectedProductId}/>
      ))}
    </List>
  );
};

export {AuctionList};