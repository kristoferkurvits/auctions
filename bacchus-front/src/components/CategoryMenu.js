import React from 'react'
import { Dropdown } from 'semantic-ui-react'


class CategoryMenu extends React.Component{
    constructor(props){
        super(props);
        this.state = {selected: ""};
        this.handleSelection = this.handleSelection.bind(this);
    }

    handleSelection = (e, data) => {
        this.props.onDropdownClick(data);
        this.setState = ({selected: data.text});
    }

    getUniqueCategories = (arr, index) => {

        const unique = arr
             .map(e => e[index])
      
             // store the keys of the unique objects
             .map((e, i, final) => final.indexOf(e) === i && i)
      
             // eliminate the dead keys & store unique objects
            .filter(e => arr[e]).map(e => arr[e]);      
      
         return unique;
    };

    render(props) {
        const { value } = this.state.selected
        let dropdownitems = null;
        if(this.props.auctions.length > 0){
            let uniqueCategories = this.getUniqueCategories(this.props.auctions, "productCategory");
            
            dropdownitems = uniqueCategories.map((category) => (
                {key: category.productId, text: category.productCategory, value: category.productCategory}
           ));
            dropdownitems[dropdownitems.length-1] = {key: "All categories", text: "All categories", value: "All categories"};
        }

        return (
            <div>
                <Dropdown
                    placeholder="Select Category"
                    fluid
                    selection
                    options={dropdownitems}
                    onChange={this.handleSelection}
                    value={value}
                    >
                    
                </Dropdown>
            </div>
            
        );
    
    };
}

    
export {CategoryMenu};
