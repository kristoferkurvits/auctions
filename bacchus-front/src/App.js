import React from "react";
import { getAuctions } from "./api"
import { Container, Header } from "semantic-ui-react";
import { AuctionList } from "./components/AuctionList"
import { CategoryMenu } from "./components/CategoryMenu"
class App extends React.Component {

  constructor(props) {
    super(props);

    this.handleDropdownClick = this.handleDropdownClick.bind(this);

    this.state = {
      auctions: [],
      apiError: "",
      selectedAuctions: [],
      selectedProductId: null
    };
  };

  async componentDidMount(){
    try {
      const response = await getAuctions();
      this.setState({auctions: response});
    } catch (error) {
      this.setState({ apiError: "Could not find any auctions" });
    }
  }


  auctionsExists(auctions){
    return auctions.length > 0;
  }

  async handleDropdownClick(data){

    let auctions = [...this.state.auctions];

    if (data.value !== "All categories"){
      auctions = auctions.filter((elem) => {
        return elem.productCategory === data.value;
      });
      this.setState({selectedAuctions: auctions});
      
    } else{
        try {
          const response = await getAuctions();
          this.setState({auctions: response, selectedAuctions: []});
        } catch (error) {
          this.setState({ apiError: "Could not find any auctions" });
        }
    }

  }

  selectedAuctionItemHandler = (productId) =>{
    this.setState({selectedProductId: productId});
  }




  render() {
    const { auctions, apiError, selectedAuctions, selectedProductId } = this.state;
    const renderAuctions = () => {

      if(this.auctionsExists(auctions)){
        return <React.Fragment>
          <CategoryMenu auctions={auctions} onDropdownClick={this.handleDropdownClick}/>
          {this.state.selectedAuctions.length
           ?
           <AuctionList auctions={selectedAuctions} 
           selectedAuctionItemHandler={this.selectedAuctionItemHandler} 
           selectedProductId={selectedProductId} />
            :
            <AuctionList auctions={auctions}
             selectedAuctionItemHandler={this.selectedAuctionItemHandler}
              selectedProductId={selectedProductId}/>}
          
        </React.Fragment>
        
      };

    };
    return (
      <Container>
        <Header as="h2" style={{ textAlign: "center", margin: 20 }}>
          Auctions
        </Header>

        {renderAuctions()}
        
        {apiError && <p>Could not fetch any articles. Please try again.</p>}
        
      </Container>
    );
  }
}

export default App;