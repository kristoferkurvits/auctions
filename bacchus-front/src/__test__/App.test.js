import { shallow, mount } from "enzyme";
import React, { Component } from 'react';
import App from "../App";
import { setupTest } from "../setupTest";
import { CategoryMenu } from "../components/CategoryMenu";
import renderer from 'react-test-renderer';
import { Form } from 'semantic-ui-react';



describe("App component and its children render", () => {
    it("renders App component without crashing", () => {
        shallow(<App />);
    
    });

    it("renders App correctly", () => {
        const tree = renderer
        .create(<App />)
        .toJSON();

        expect(tree).toMatchSnapshot();
      });
      
    it("renders entire application", () => {
        mount(<App />);
    });
})




describe("Dropdown menu", () => {
    const auction = {
        "productId": "036d9cc3-b0a3-442c-96ac-4291b3ac2894",
        "productName": "MSI GeForce GTX 1080",
        "productDescription": "Average MSI GeForce GTX 1080",
        "productCategory": "Computer hardware",
        "biddingEndDate": "2020-09-21T11:00:07Z"
    };
    it("Data appears in dropdown from props", () => {
        const wrapper = mount(<CategoryMenu auctions={auction} />);
        const value = wrapper.find(CategoryMenu).props().auctions["productName"];

        expect(value).toBe("MSI GeForce GTX 1080");


    });

    it("Dropdown handler works", () => {
        const value = "value";
        const mockMyEventHandler = jest.fn();
        const wrapper = mount(<CategoryMenu auctions={auction} onChange={mockMyEventHandler}/>);
        wrapper.prop("onChange")(value);

        expect(mockMyEventHandler).toHaveBeenCalledWith(value);
        
    });

    

});



