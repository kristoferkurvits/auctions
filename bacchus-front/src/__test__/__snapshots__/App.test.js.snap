// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component and its children render renders App correctly 1`] = `
<div
  className="ui container"
>
  <h2
    className="ui header"
    style={
      Object {
        "margin": 20,
        "textAlign": "center",
      }
    }
  >
    Auctions
  </h2>
  
</div>
`;
